added aep api with request and response conversions

jasonwaters opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

This PR implements Delivery to Konductor Request conversion and Konductor to Delivery Response conversion.  It covers the basics, although there will be further testing and additions in subsequent PRs.

You will notice a new `src/api` folder that contains entry points for all the apis we currently support (delivery and decisioning engine), and a new one for aep.  I found this to be the most efficient way to introduce aep as a new API without code duplication.  The SDK continues to run with no major modifications, the only change is which api is used to fulfill delivery requests.

Conversion methods are in `target-tools` so that they may be re-used elsewhere (like in at.js) as we see fit.

For rapid development/testing you may find it useful to use the `scratch.spec.js` test.  Be sure to remove the `skip` on the describe function first 😉️.  The test instantiates two SDK instances.  One with an `edgeConfigId` and one without.  Then is sends the same Delivery request to each instance `getOffers`.  Finally, it compares the result of each to see how close they are.  It's a good way to see the gaps and test various requests.

When running this code locally, you will need to ensure `aep-edge-nodejs-sdk` is cloned and built.  Packages in this repo look for that code in `~/Development/aep-edge-nodejs-sdk`.  I recommend cloning it there or updating the dependencies to point to the location you have it cloned.  This is only temporary until we do an official release to npm.  Adding the dependency with a file system reference avoids the hassle of manually doing `npm link`, and npm will do it for you creating symlinks as needed.

## Related Issues

- [Implement Delivery->Konductor request conversion](https://jira.corp.adobe.com/browse/TNT-41425)
- [Implement Konductor->Delivery response conversion](https://jira.corp.adobe.com/browse/TNT-41426)
- [Implement AEP service](https://jira.corp.adobe.com/browse/TNT-41917)

## How Has This Been Tested?

Unit tests

XDex commented on packages/target-nodejs-sdk/test/api.spec.js:
```diff
@@ -0,0 +1,303 @@
+      requestId: "244f07a4895a47ef94a003bfb578d773",
+      context: {
+        channel: "web",
+        timeOffsetInMinutes: -360
```
Without this fix the test fails on my machine:
```suggestion
        timeOffsetInMinutes: expect.any(Number)
```

jasonwaters replied:
Good catch, forgot to change that

XDex commented on packages/target-tools/src/transforms/transformUtils.js:
```diff
@@ -0,0 +1,267 @@
+const METRIC_ACTION_PROPERTY_MAP = [
+  { aep: "type", target: "type" },
+  { aep: "selector", target: "selector" },
+  { aep: "???", target: "eventToken" }
```
```suggestion
  { aep: "???", target: "eventToken" } // placeholder until support for unencrypted eventTokens (scopeDetails) is implemented in DeliveryAPI
```

XDex commented on packages/target-tools/src/transforms/transformUtils.js:
```diff
@@ -0,0 +1,267 @@
+
+export function isFormBasedActivity(data) {
+  const { selector } = data;
+  return selector === HEAD;
```
we shouldn't determine the action as Form-based based on HEAD selector, as a HEAD selector can also be specified in a VEC action (by editing the selector in the UI)
I'd suggest differentiating FBC vs VEC only based on item schema

XDex commented on packages/target-tools/src/transforms/requestTransform.js:
```diff
@@ -0,0 +1,438 @@
+
+  const newIdentityMap = { ...identityMap };
+
+  if (isDefined(marketingCloudVisitorId)) {
```
extract method

XDex commented on packages/target-tools/src/transforms/requestTransform.js:
```diff
@@ -0,0 +1,438 @@
+    );
+  }
+
+  if (isDefined(customerIds) && isArray(customerIds)) {
```
extract method

XDex commented on packages/target-tools/src/transforms/requestTransform.js:
```diff
@@ -0,0 +1,438 @@
+      targetParams = { ...parameters };
+      targetProfileParams = { ...profileParameters };
+
+      const xdm = {
```
extract method

XDex commented on packages/target-tools/src/transforms/requestTransform.js:
```diff
@@ -0,0 +1,438 @@
+      const xdm = {
+        timestamp: new Date(),
+        implementationDetails: {
+          name: "https://ns.adobe.com/experience/aep-edge-nodejs-sdk",
```
extract constant

XDex commented on packages/target-tools/src/transforms/requestTransform.js:
```diff
@@ -0,0 +1,438 @@
+        timestamp: new Date(),
+        implementationDetails: {
+          name: "https://ns.adobe.com/experience/aep-edge-nodejs-sdk",
+          version: "1.0.0",
```
import from `package.json` instead of hardcoding

XDex commented on packages/target-tools/src/transforms/requestTransform.js:
```diff
@@ -0,0 +1,438 @@
+    entries: [...entries]
+  };
+
+  if (isNotBlank(imsOrgId) && isNotBlank(sessionId)) {
```
extract method

jasonwaters replied:
what are you trying to achieve by extracting a method here?  in general, i prefer to extract methods for readability or code reuse.  this logic check is only used in this one place.  is it hard to read?

XDex replied:
Just trying to improve readability and reduce method length by abstracting away chunks of code that can be isolated in separate methods. By replacing these 6 lines of code with a simple call to `setSessionCookie` the method becomes more readable.

XDex commented on packages/target-tools/src/transforms/requestTransform.js:
```diff
@@ -0,0 +1,438 @@
+    });
+  }
+
+  if (isNotBlank(konductorIdentity)) {
```
extract method

XDex commented on packages/target-tools/src/transforms/responseTransform.js:
```diff
@@ -0,0 +1,233 @@
+      const { activity = {} } = scopeDetails;
+      const { id: activityId } = activity;
+
+      if (PAGE_WIDE_SCOPE === scope) {
```
extract method

XDex commented on packages/target-tools/src/transforms/responseTransform.js:
```diff
@@ -0,0 +1,233 @@
+      const mboxItems = items.filter(isMboxSchema);
+      const viewItems = items.filter(isViewSchema);
+
+      if (isNonEmptyArray(mboxItems)) {
```
extract method

XDex commented on packages/target-tools/src/transforms/responseTransform.js:
```diff
@@ -0,0 +1,233 @@
+        );
+      }
+
+      if (isNonEmptyArray(viewItems)) {
```
extract method

XDex reviewed (commented):
@jasonwaters Looks great! Please take a look at some comments

jasonwaters commented on packages/target-nodejs-sdk/src/api/aep.js:
```diff
@@ -0,0 +1,72 @@
+      imsOrgId,
+      sessionId,
+      deliveryRequest,
+      atjsVersion,
```
i agree with the sentiment, but let's hold off and do it as a separate PR.  I believe, the `version` sent to delivery API is strictly the atjs version.  And it is just a number.  So if we change it to `sdk version` we should also introduce some other qualifier for the sdk name.  Thoughts?

XDex replied:
>  I believe, the version sent to delivery API is strictly the atjs version

I'm probably missing something here, but why would we want to send an at.js version from the Node.JS SDK? The version sent to Delivery API should always refer to the SDK it was sent from.
I suggest removing this version param altogether and just set SDK name and version in `xdm.implementationDetails` for AEP and in `X-EXC-SDK` headers for direct Delivery the way it's currently implemented in `helper.js/createHeaders()`

jasonwaters commented on packages/target-nodejs-sdk/src/api/index.js:
```diff
@@ -0,0 +1,88 @@
+    return createAepApi(sdkConfig, configuration);
+  }
+
+  return useBeacon && isBeaconSupported()
```
We support it because the node.js sdk also produces a browser build that some people use instead of at.js.  In that case it leverages beacon for notifications.

XDex replied:
Ok, I think we should implement a way to detect when the browser build is used and change the SDK name header to `AdobeTargetNodeBrowser` in this case, in a separate task.

jasonwaters commented:
@XDex thanks for your review, i really appreciate you taking the time.  I've made some tweaks per your comments.  If you don't mind, i'd like to go ahead and merge this PR and take care of extracting methods in subsequent PRs, where it makes sense.

XDex commented on packages/target-nodejs-sdk/src/api/delivery.js:
```diff
@@ -0,0 +1,70 @@
+ */
+export function createBeaconDeliveryApi(configuration) {
+  return {
+    execute: (organizationId, sessionId, deliveryRequest, atjsVersion) => {
```
as mentioned above, let's remove version altogether

XDex commented on packages/target-nodejs-sdk/src/api/delivery.js:
```diff
@@ -0,0 +1,75 @@
+        sessionId
+      };
+
+      if (isDefined(configuration.version)) {
```
as mentioned above, let's remove version altogether, as SDK name and version will be sent in `X-EXC-SDK*` headers

XDex reviewed (approved):
@jasonwaters Sure, I'm ok with merging this one and addressing the remaining comments in subsequent PRs, please open separate tasks for tracking this.
Thanks!

jasonwaters commented:
> @jasonwaters Sure, I'm ok with merging this one and addressing the remaining comments in subsequent PRs, please open separate tasks for tracking this.
> Thanks!

https://jira.corp.adobe.com/browse/TNT-42323
